keanu-project/src/main/java/io/improbable/keanu/algorithms/sampling/Forward.java

```diff
@@ -0,0 +1,131 @@
+package io.improbable.keanu.algorithms.sampling;
+
+import com.google.common.collect.Sets;
+import io.improbable.keanu.KeanuRandom;
+import io.improbable.keanu.algorithms.NetworkSamples;
+import io.improbable.keanu.algorithms.PosteriorSamplingAlgorithm;
+import io.improbable.keanu.algorithms.ProbabilisticModel;
+import io.improbable.keanu.algorithms.Variable;
+import io.improbable.keanu.algorithms.graphtraversal.TopologicalSort;
+import io.improbable.keanu.algorithms.mcmc.NetworkSamplesGenerator;
+import io.improbable.keanu.algorithms.mcmc.SamplingAlgorithm;
+import io.improbable.keanu.network.BayesianNetwork;
+import io.improbable.keanu.network.LambdaSection;
+import io.improbable.keanu.network.TransitiveClosure;
+import io.improbable.keanu.util.status.StatusBar;
+import io.improbable.keanu.vertices.Vertex;
+import org.nd4j.base.Preconditions;
+
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+public class Forward implements PosteriorSamplingAlgorithm {
+
+    private final KeanuRandom random;
+    private final boolean calculateSampleProbability;
+
+    public static ForwardBuilder builder() {
+        return new ForwardBuilder();
+    }
+
+    public Forward(KeanuRandom random, boolean calculateSampleProbability) {
+        this.random = random;
+        this.calculateSampleProbability = calculateSampleProbability;
+    }
+
+    /**
+     * Samples from the prior of a Probabilistic Model.
+     *
+     * Samples are taken by sampling from the prior of the desired variables in topological order.
+     *
+     * @param model the model to sample from
+     * @param variablesToSampleFrom the variables to sample from
+     * @param sampleCount the number of samples to take
+     * @return sampling samples of a computable graph
+     */
+    @Override
+    public NetworkSamples getPosteriorSamples(ProbabilisticModel model, List<? extends Variable> variablesToSampleFrom, int sampleCount) {
+        return generatePosteriorSamples(model, variablesToSampleFrom)
+            .generate(sampleCount);
+    }
+
+    @Override
+    public NetworkSamplesGenerator generatePosteriorSamples(ProbabilisticModel model, List<? extends Variable> variablesToSampleFrom) {
+        return new NetworkSamplesGenerator(setupSampler(model, variablesToSampleFrom), StatusBar::new);
+    }
+
+    private SamplingAlgorithm setupSampler(ProbabilisticModel model, List<? extends Variable> variablesToSampleFrom) {
+        List<? extends Variable> latentVariables = model.getLatentVariables();
+        Preconditions.checkArgument(latentVariables.size() > 0, "Your model must contain latent variables in order to forward sample.");
+
+        for (Variable variable : variablesToSampleFrom) {
+            Preconditions.checkArgument(variable instanceof Vertex, "The Forward Sampler only works for Variables of type Vertex. Received : " + variable);
+        }
```

Comment:
I think it would be cleaner to cast all your `Variable`s to `Vertex` here, just once, instead of multiple times further down.